common/scala/src/main/scala/whisk/common/tracing/OpenTracingProvider.scala

```diff
@@ -0,0 +1,204 @@
+ */
+
+package whisk.common.tracing
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+import brave.Tracing
+import brave.opentracing.BraveTracer
+import brave.sampler.Sampler
+import io.opentracing.{Span, SpanContext, Tracer}
+import io.opentracing.util.GlobalTracer
+import io.opentracing.propagation.{Format, TextMapExtractAdapter, TextMapInjectAdapter}
+import zipkin2.reporter.{AsyncReporter, Sender}
+import zipkin2.reporter.okhttp3.OkHttpSender
+import pureconfig._
+import whisk.common.{LogMarkerToken, TransactionId}
+import whisk.core.ConfigKeys
+
+import scala.ref.WeakReference
+
+/**
+ * OpenTracing based implementation for tracing
+ */
+class OpenTracer(val tracer: Tracer) extends WhiskTracer {
+
+  /**
+   * Start a Trace for given service.
+   *
+   * @param transactionId transactionId to which this Trace belongs.
+   * @return TracedRequest which provides details about current service being traced.
+   */
+  override def startSpan(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {
+
+    //initialize list for this transactionId
+    val spanList: List[WeakReference[Span]] = TracingCacheProvider.spanMap.get(transactionId.meta.id).getOrElse(Nil)
+
+    val activeSpan: Option[Span] = spanList match {
+      case Nil => {
+        val spanBuilder = tracer
+          .buildSpan(logMarker.action)
+          .withTag("transactionId", transactionId.meta.id)
+
+        Some(
+          TracingCacheProvider.contextMap
+            .get(transactionId.meta.id)
+            .map(spanBuilder.asChildOf(_).startActive(true).span())
+            .getOrElse(spanBuilder.ignoreActiveSpan().startActive(true).span()))
+      }
+      case _ =>
+        Some(
+          tracer
+            .buildSpan(logMarker.action)
+            .withTag("transactionId", transactionId.meta.id)
+            .asChildOf(spanList(0).get.get)
+            .startActive(true)
+            .span())
+
+    }
+    //add active span to list
+    if (activeSpan.isDefined)
+      TracingCacheProvider.spanMap.put(transactionId.meta.id, spanList.::(new WeakReference(activeSpan.get)))
+  }
+
+  /**
+   * Finish a Trace associated with given transactionId.
+   *
+   * @param transactionId
+   */
+  override def finishSpan(transactionId: TransactionId): Unit = {
+    clear(transactionId)
+  }
+
+  /**
+   * Register error
+   *
+   * @param transactionId
+   */
+  override def error(transactionId: TransactionId): Unit = {
+    clear(transactionId)
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  override def getTraceContext(transactionId: TransactionId): Option[Map[String, String]] = {
+    var contextMap: Option[Map[String, String]] = None
+    TracingCacheProvider.spanMap.get(transactionId.meta.id) match {
+      case Some(spanList) => {
+        var map: java.util.Map[String, String] = new java.util.HashMap()
```

Comment:
Mutability needed? Same for the above. You could also use a Scala Map by default and call `asJava` when passing into javaland rather than doing the inverse.